pkg/agent/openflow/cookie/allocator.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2019 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cookie
+
+import (
+	"fmt"
+)
+
+const (
+	BitwidthRound    = 16
+	BitwidthCategory = 8
+	BitwidthReserved = 64 - BitwidthCategory - BitwidthRound
+
+	RoundMask    = uint64(0xffff000000000000)
+	CategoryMask = uint64(0x0000ff0000000000)
+)
+
+// Category represents the flow entry category.
+type Category uint64
+
+const (
+	Default Category = iota
+	Gateway
+	Node
+	Pod
+	Service
+	Policy
+)
+
+func (c Category) String() string {
+	switch c {
+	case Default:
+		return "Default"
+	case Gateway:
+		return "Gateway"
+	case Node:
+		return "Node"
+	case Pod:
+		return "Pod"
+	case Service:
+		return "Service"
+	case Policy:
+		return "Policy"
+	default:
+		return "Invalid"
+	}
+}
+
+// ID defines segments a cookie ID contains. An ID is composed like:
+//  |------------------------- ID --------------------------|
+//  |- round 16bits -|- category 8bits -|- reserved 40bits -|
+// The round segment represents the round id.
+// The category segment represents the category of flow this ID belongs.
+type ID uint64
+
+func newID(round uint64, cat Category) ID {
+	r := uint64(0)
+	r |= round << (64 - BitwidthRound)
+	r |= (uint64(cat) << BitwidthReserved) & CategoryMask
+	return ID(r)
+}
+
+// Raw returns the unit64 type value of the ID.
+func (i ID) Raw() uint64 {
+	return uint64(i)
+}
+
+// Round returns the round number of the ID.
+func (i ID) Round() uint64 {
+	return i.Raw() >> (64 - BitwidthRound)
+}
+
+// Category returns the category of the ID.
+func (i ID) Category() Category {
+	return Category((i.Raw() & CategoryMask) >> BitwidthReserved)
+}
+
+// String returns the string representation of the ID.
+func (i ID) String() string {
+	return fmt.Sprintf("<round:%d,category:%s>", i.Round(), i.Category().String())
+}
+
+// Allocator defines operations of a cookie ID allocator.
+type Allocator interface {
+	// Request cookie IDs of flow categories.
+	Request(cat Category) ID
+}
+
+type allocator struct {
+	round uint64
+}
+
+// Request returns a ID with the given category.
+func (a *allocator) Request(cat Category) ID {
+	return newID(a.round, cat)
```

Comment:
Do we expect `round` changes at runtime? If not, is it worth to calculate the known IDs repeatedly for every flow?

Reply:
They would not be changed at current time. I changed to cache these IDs instead create new one each time.

Reply:
I'm not sure why we are caching them when they are so cheap to compute. Won't that be a problem to handle reconnects, @wenyingd ?

Reply:
If we want to resync the flows after a reconnect, including replay the Openflow entries for existing resources, and delete stale flows, two seperate round IDs are needed, one is used as the mark of the new round, and the other is the last round. And the workflow should be replay flows in the new round firstly, and then delete those in last round secondly.  

In this case, if we have two seprate id-allocator, one for the last round, and the other for current, I think both caching the IDs or recalculate ID would be OK.

Reply:
Just personal opinion to make this as simple as possible, feel free to ignore if it doesn't make sense to you:
Perhaps it's not "caching", but getting known values, they are just several static numbers, I understand it's dynamic allocated in the first version when each flow has its own ID, but it has become static now, does it still make sense to re-calculate them and uses the terminologies "allocate", "request"? I'm even thinking the values can just be variables and directly used when building flows, as it makes no difference for clients that `cookie.Default` is a category ID or a cookieID for this category. Even the round num can change at runtime, it could just refresh the variables in one shot.

Reply:
either way is fine with me, I don't have a strong preference. As far as the "interface" is concerned (i.e. how we get them from the "allocator" to the client), I think it may make sense to stick with methods, as we may want to encode additional flow context in the cookie in the future. Regardless, I would agree with @tnqn that the "allocator" terminology does not make too much sense any more. It's more of a cookie "builder", even though as you pointed out we are not building anything dynamically at the moment.

Reply:
OK, I will keep current implementation and add a todo to the `preallocate` map.